runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/DoFnOp.java

```diff
@@ -237,9 +292,39 @@ private String getTimerStateId(DoFnSignature signature) {
     return builder.toString();
   }
 
+  private void attemptStartBundle() {
+    if (isBundleStarted.compareAndSet(false, true)) {
+      currentBundleElementCount.set(0L);
+      bundleStartTime.set(System.currentTimeMillis());
+      pushbackFnRunner.startBundle();
```

Comment:
Shall we invoke scheduleNextBundleCheck() after startBundle()? Right now schedule() is invoked in multiple places. Seems we only need to schedule the check here all the time.

Reply:
There are two models. 1) the current model is to regularly check in the background to see whether max bundling time is hit; this is only roughly accurate but more efficient 2) schedule a timer for EVERY bundle start. This is more accurate but we may end up scheduling too many timer when the QPS is too high. Let's discuss this